### PR TITLE
Save groupby codes after factorizing, pass to flox

### LINKED ISSUE #7601: groupby_bins groups not correctly applied with built-in methods
### Setup
I want to calculate image statistics per chunk in one dimension. Let's assume a very small image for demonstration purposes:

```python
a = xr.DataArray(np.arange(12).reshape(6,2), dims=('x', 'y'))
a
```
image
Trying to chunk this into three subimages, I use these bins into the x dimension:
```python
x_bins = (0, 2, 4, 6)
```
I look at the groups this creates by default:
```python
for iv, g in a.groupby_bins('x', x_bins):
    print(iv)
    print(g)
```
image
I don't understand the use-case for this grouping, as it's missing the beginning and is having uneven sized last group (Obviously a follow-error from not including the first row).

To force the even chunking of the image I need to call it with these parameters:
```python
groups = a.groupby_bins('x', x_bins, include_lowest=True, right=False)
for iv, g in groups:
    print(iv)
    print(g)
```
image

### Issue
But now, calculating the mean value of each group, I get different results when doing it by hand using the groups or doing it using the groups inherent method `mean()`:

image

Indeed, I verified, that these results are what one gets, using the first version of applying the bins:
image

The same is true when I use the elliptical operator to receive the mean over the remaining dimensions (note, the 2nd cell here is using the `groups` variable as defined in the cell before, so should really return the same values, but it doesn't:
image

### Application
I believe that `groupby_bins` is the most appropriate tool to do this in xarray. I wished that one could enforce the dask-chunks in dask arrays to survive and return stats from them, but haven't found a way to do that.

### What did you expect to happen?

That the inherent stats methods of the `groups` method respect the interval use constraints from the `groupby_bins` call.

I also have verified that the same problem exists with `groups.std()`.

### Minimal Complete Verifiable Example

```Python
import xarray as xr
import numpy as np

a = xr.DataArray(np.arange(12).reshape(6,2), dims=('x', 'y'))

x_bins = (0, 2, 4, 6)

default_groups = a.groupby_bins('x', x_bins)
my_groups = a.groupby_bins('x', x_bins, include_lowest=True, right=False)

print("Weird grouping using default call:")
for iv, g in default_groups:
    print("Interval:",iv)
    print(g.data)
    print()
    
print("Evenly chunked using `my_groups`:")    
for iv, g in my_groups:
    print("Interval:", iv)
    print(g.data)
    print()

print("Calculating mean on my own using loop over groups:")
for iv, g in my_groups:
    print(g.mean('x').data)
    
print("Calculting same using my_groups.mean()")
print("No dim given:")
print(my_groups.mean().data.T)
print("using mean('x'):")
print(my_groups.mean('x').data.T)

print("These results come from the default groups!:")
for iv, g in default_groups:
    print(g.mean('x').data)
    
print("STD has the same issue")
```

### MVCE confirmation

- [X] Minimal example — the example is as focused as reasonably possible to demonstrate the underlying issue in xarray.
- [X] Complete example — the example is self-contained, including all data and the text of any traceback.
- [X] Verifiable example — the example copy & pastes into an IPython prompt or [Binder notebook](https://mybinder.org/v2/gh/pydata/xarray/main?urlpath=lab/tree/doc/examples/blank_template.ipynb), returning the result.
- [X] New issue — a search of GitHub Issues suggests this is not a duplicate.

### Relevant log output

_No response_

### Anything else we need to know?

_No response_

### Environment

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.10.9 | packaged by conda-forge | (main, Feb  2 2023, 20:20:04) [GCC 11.3.0]
python-bits: 64
OS: Linux
OS-release: 6.0.12-76060006-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: en_US.UTF-8
LANG: en_US.UTF-8
LOCALE: ('en_US', 'UTF-8')
libhdf5: 1.12.2
libnetcdf: 4.9.1

xarray: 2023.2.0
pandas: 1.5.3
numpy: 1.23.5
scipy: 1.10.1
netCDF4: 1.6.3
pydap: None
h5netcdf: None
h5py: 3.8.0
Nio: None
zarr: None
cftime: 1.6.2
nc_time_axis: None
PseudoNetCDF: None
rasterio: 1.3.6
cfgrib: None
iris: None
bottleneck: 1.3.7
dask: 2023.3.0
distributed: 2023.3.0
matplotlib: 3.7.1
cartopy: 0.21.1
seaborn: 0.12.2
numbagg: None
fsspec: 2023.3.0
cupy: None
pint: None
sparse: None
flox: 0.6.8
numpy_groupies: 0.9.20
setuptools: 67.5.1
pip: 23.0.1
conda: installed
pytest: 7.1.3
mypy: None
IPython: 8.7.0
sphinx: None

</details>

### PR DESCRIPTION
**Pull request by dcherian**
- [x] Closes #6689,
- xref #6610
- [x] Closes #7601 
- [x] Tests added

This is an alternative to #6689. 
- There I tried to avoid factorizing in the GroupBy constructor, and passed the `by` variable directly to `flox`. Most GroupBy methods however depend on various steps in `__init__`, so it became messy.
- Here I instead preserve factorizing in the constructor and pass the factorized codes to `flox`. This simplifies things a lot. Since we'll want to preserve the "for loop over groups" approach for `GroupBy.map`, we'll need something like this anyway.
 
The large amount of deleted code in `_flox_reduce` here suggests to me that this is the better approach.

I think we could also use this to generalize to multiple groupers:
- factorize each,
- `ravel_muti_index` to generate a singe variable to groupby
- apply
- reshape to output shape
- assign new indices.

### CODE REVIEW

File: xarray/core/groupby.py [hunk 1]
```diff
@@ -388,63 +483,20 @@ def __init__(
                 "dimension"
             )
 
-        full_index = None
```
Comment (dcherian) [inline xarray/core/groupby.py:391]: Just refactored with no functionality changes except to return `codes`

**Comment by Illviljan** [comment]
Hmm, did I mess something up? I believe I only changed typing related things.

_bins suffix seems to have disappeared:
```
     def test_groupby_bins_multidim(self):
        array = self.make_groupby_multidim_example_array()
        bins = [0, 15, 20]
        bin_coords = pd.cut(array["lat"].values.flat, bins).categories
        expected = DataArray([16, 40], dims="lat_bins", coords={"lat_bins": bin_coords})
        actual = array.groupby_bins("lat", bins).map(lambda x: x.sum())
>       assert_identical(expected, actual)
E       AssertionError: Left and right DataArray objects are not identical
E       Differing dimensions:
E           (lat_bins: 2) != (lat: 2)
E       Coordinates only on the left object:
E         * lat_bins  (lat_bins) object (0, 15] (15, 20]
E       Coordinates only on the right object:
E         * lat       (lat) object (0, 15] (15, 20]
```

File: xarray/core/groupby.py [hunk 2]
```diff
@@ -45,6 +45,9 @@
 
     GroupKey = Any
 
+    T_GroupIndicesListInt = list[list[int]]
+    T_GroupIndices = Union[T_GroupIndicesListInt, list[slice], np.ndarray]
```
Comment (Illviljan) [inline xarray/core/groupby.py:49]: Would be nice to reduce this to 1 consistent type.

File: xarray/core/groupby.py [hunk 3]
```diff
@@ -209,6 +217,9 @@ def __getitem__(self, key):
             key = key[0]
         return self.values[key]
 
+    def copy(self, deep: bool = True, data: Any = None):
+        raise NotImplementedError
```
Comment (Illviljan) [inline xarray/core/groupby.py:221]: I (and mypy) don't understand how `_DummyGroup` manages to get through the GroupBy init without crashing?

**Review by Illviljan — APPROVED** [review]
Thanks, @dcherian. This LGTM. :)

File: xarray/core/groupby.py (hunk 2 above)
Comment (dcherian) [reply xarray/core/groupby.py:49]: The `np.ndarray` one is potentially removable after https://github.com/pydata/xarray/issues/2157

**Pull request merged**